File: mailblastr-ruby/lib/mailblastr/automations.rb
# frozen_string_literal: true

module Mailblastr
  # Automations are DOMAIN-FIRST: `domain` is required on create, and only
  # Events.send calls naming the same domain trigger them.
  module Automations
    class << self
      # POST /automations
      #   Mailblastr::Automations.create({ name: "Welcome series", domain: "yourdomain.com",
      #                                    trigger: "contact.created" })
      # The built-in "mailblastr:schedule" trigger fires once at
      # `trigger_config` ({ at: ISO 8601 instant, timezone: IANA name }),
      # enrolling every contact of the domain's pool — `trigger_config` is
      # required with that trigger and not accepted on any other.
      def create(params)
        Client.require_domain!(params, "Automations.create")
        Client.request(:post, "/automations", body: params)
      end

      # GET /automations/:id
      def get(automation_id)
        Client.request(:get, "/automations/#{Client.path_escape(automation_id)}")
      end

      # GET /automations
      def list(params = {})
        Client.request(:get, "/automations", query: Client.pagination(params))
      end

      # PATCH /automations/:id — params: { name:, status: "enabled"|"disabled", ... }
      # `trigger_config` ({ at:, timezone: }) updates the "mailblastr:schedule"
      # trigger's schedule (only valid on automations with that trigger).
      def update(automation_id, params)
        Client.request(:patch, "/automations/#{Client.path_escape(automation_id)}", body: params)
      end

      # Append a step. POST /automations/:id/steps — params: { type:, config:, key: }
      # The automation must be disabled first, and `type: "trigger"` is
      # rejected here (the trigger lives on the automation, not in `steps`).
      def add_step(automation_id, params)
        Client.request(:post, "/automations/#{Client.path_escape(automation_id)}/steps", body: params)
      end

      # Edit a step in place (automation must be disabled).
      # PATCH /automations/:id/steps/:step_id
      def update_step(automation_id, step_id, params)
        Client.request(
          :patch,
          "/automations/#{Client.path_escape(automation_id)}/steps/#{Client.path_escape(step_id)}",
          body: params
        )
      end

      # Delete a step. DELETE /automations/:id/steps/:step_id
      def delete_step(automation_id, step_id)
        Client.request(:delete, "/automations/#{Client.path_escape(automation_id)}/steps/#{Client.path_escape(step_id)}")
      end

      # Build (or extend) the automation's steps from a prompt.
      # POST /automations/:id/ai — params: { prompt:, template_ids:, events:, attach: }
      # `prompt` is required and capped at 2000 characters. Without `attach` the
      # automation must have no steps yet; pass `attach` ({ from:, type:,
      # before: }) to append to an existing graph. The automation must be
      # stopped, and the route is limited to 20 requests per minute per account.
      def create_with_ai(automation_id, params)
        Client.request(:post, "/automations/#{Client.path_escape(automation_id)}/ai", body: params)
      end

      # List an automation's runs. `status` filters to specific run statuses
      # ("running", "completed", "failed", "skipped") and accepts an Array or a
      # comma-separated String. GET /automations/:id/runs
      def runs(automation_id, params = {})
        query = Client.pagination(params)
        status = Client.opt(params, :status)
        query[:status] = status.is_a?(Array) ? status.join(",") : status unless status.nil?
        Client.request(:get, "/automations/#{Client.path_escape(automation_id)}/runs", query: query)
      end

      # Retrieve a single run with its step trace. GET /automations/:id/runs/:run_id
      def get_run(automation_id, run_id)
        Client.request(:get, "/automations/#{Client.path_escape(automation_id)}/runs/#{Client.path_escape(run_id)}")
      end

      # Stop an automation — no new runs; in-progress runs finish. POST /automations/:id/stop
      def stop(automation_id)
        Client.request(:post, "/automations/#{Client.path_escape(automation_id)}/stop")
      end

      # DELETE /automations/:id
      def delete(automation_id)
        Client.request(:delete, "/automations/#{Client.path_escape(automation_id)}")
      end
    end
  end
end

File: mailblastr-php/src/Resources/Automations.php
<?php

declare(strict_types=1);

namespace Mailblastr\Resources;

use Mailblastr\Client;

/**
 * Automations — multi-step workflows triggered by events.
 * DOMAIN-FIRST: 'domain' is REQUIRED on create; only `events->send()` calls
 * with the same 'domain' trigger the automation, so the same event name across
 * several products can never trigger the wrong one.
 */
class Automations extends Resource
{
    /**
     * Create an automation. POST /automations
     *
     * @param array $payload ['name' => …, 'domain' => REQUIRED,
     *                       'trigger' => 'contact.created' | 'mailblastr:schedule' |
     *                                    'email.opened' | 'email.clicked' |
     *                                    'email.replied' | 'email.bounced' |
     *                                    'email.delivered' | any custom event name,
     *                       'trigger_config' => ['at' => ISO 8601 instant,
     *                                            'timezone' => IANA name]
     *                                           (required with the
     *                                           'mailblastr:schedule' trigger;
     *                                           not accepted on any other),
     *                       'status' => 'enabled'|'disabled' (default 'disabled'),
     *                       'steps' => [['key' => …, 'type' => …, 'config' => …], …],
     *                       'connections' => [['from' => …, 'to' => …, 'type' => …], …]]
     */
    public function create(array $payload): array
    {
        return $this->client->request('POST', '/automations', $payload);
    }

    /** Retrieve an automation (steps, connections, enrollments). GET /automations/:id */
    public function get(string $id): array
    {
        return $this->client->request('GET', '/automations/' . Client::e($id));
    }

    /**
     * List automations. GET /automations
     *
     * @param array $params Cursor pagination: limit, after, before.
     */
    public function list(array $params = []): array
    {
        return $this->client->request('GET', '/automations' . $this->paginationQuery($params));
    }

    /**
     * Update an automation (name/status/domain/trigger_config/connections).
     * 'trigger_config' (['at' => …, 'timezone' => …]) updates the
     * 'mailblastr:schedule' trigger's schedule (only valid on automations with
     * that trigger). PATCH /automations/:id
     */
    public function update(string $id, array $payload): array
    {
        return $this->client->request('PATCH', '/automations/' . Client::e($id), $payload);
    }

    /**
     * Append a step to an automation. The automation must be disabled first
     * (stop it with {@see stop()}). POST /automations/:id/steps
     *
     * @param array $payload ['type' => 'delay'|'send_email'|'wait_for_event'|
     *                                  'condition'|'split'|'add_to_segment'|
     *                                  'contact_update'|'contact_delete',
     *                       'config' => …, 'key' => …]
     *                       The trigger is set on the automation, not added as
     *                       a step — type 'trigger' is a 422 here.
     */
    public function addStep(string $id, array $payload): array
    {
        return $this->client->request('POST', '/automations/' . Client::e($id) . '/steps', $payload);
    }

    /**
     * Update a step. The automation must be disabled first.
     * PATCH /automations/:id/steps/:stepId
     *
     * @param array $payload ['type' => …, 'config' => …, 'key' => …]
     */
    public function updateStep(string $id, string $stepId, array $payload): array
    {
        return $this->client->request(
            'PATCH',
            '/automations/' . Client::e($id) . '/steps/' . Client::e($stepId),
            $payload
        );
    }

    /** Delete a step from an automation. DELETE /automations/:id/steps/:stepId */
    public function deleteStep(string $id, string $stepId): array
    {
        return $this->client->request(
            'DELETE',
            '/automations/' . Client::e($id) . '/steps/' . Client::e($stepId)
        );
    }

    /**
     * Build (or extend) an automation's steps with AI. The automation must be
     * stopped. Limited to 20 requests per minute per account.
     * POST /automations/:id/ai
     *
     * @param array $payload ['prompt' => REQUIRED (max 2000 chars),
     *                       'template_ids' => [ … ] (first 10 used),
     *                       'events' => [ … ] (first 10 used),
     *                       'attach' => ['from' => trigger or step key,
     *                                    'type' => 'default'|'condition_met'|
     *                                              'condition_not_met'|
     *                                              'event_received'|'timeout',
     *                                    'before' => step key]]
     *                       Without 'attach' the automation must have no steps
     *                       yet; with it, the generated steps are spliced in.
     */
    public function ai(string $id, array $payload): array
    {
        return $this->client->request('POST', '/automations/' . Client::e($id) . '/ai', $payload);
    }

    /**
     * List an automation's runs. GET /automations/:id/runs
     *
     * @param array $params Cursor pagination (limit, after, before) plus an
     *                      optional 'status' filter — a comma-separated list of
     *                      run statuses ('running', 'completed', 'failed',
     *                      'skipped'), applied before pagination.
     */
    public function runs(string $id, array $params = []): array
    {
        return $this->client->request(
            'GET',
            '/automations/' . Client::e($id) . '/runs' . $this->paginationQuery($params, ['status'])
        );
    }

    /** Retrieve a single automation run (with its step trace). GET /automations/:id/runs/:runId */
    public function getRun(string $id, string $runId): array
    {
        return $this->client->request(
            'GET',
            '/automations/' . Client::e($id) . '/runs/' . Client::e($runId)
        );
    }

    /** Stop an automation — prevents new runs; in-progress runs finish. POST /automations/:id/stop */
    public function stop(string $id): array
    {
        return $this->client->request('POST', '/automations/' . Client::e($id) . '/stop');
    }

    /** Delete an automation. DELETE /automations/:id */
    public function remove(string $id): array
    {
        return $this->client->request('DELETE', '/automations/' . Client::e($id));
    }
}
